src/java/org/jivesoftware/openfire/session/LocalClientSession.java

```diff
@@ -932,6 +932,26 @@ public void setMessageCarbonsEnabled(boolean enabled) {
      */
     @Override
 	public boolean canProcess(Packet packet) {
+        if (JiveGlobals.getBooleanProperty(DISCARD_CARBON_SPOOF, true)) {
+            Message msg = (Message)packet;
+            if (msg != null) {
+                PacketExtension carbon = msg.getExtension("received", "urn:xmpp:carbons:2");
+                if (carbon != null) {
+                    try {
+                        Element copy = carbon.getElement().element("forwarded").element("message");
+                        JID carbonFrom = new JID(copy.attribute("from").getValue());
+                        if (carbonFrom.asBareJID() != packet.getFrom()) {
+                            Log.info("Carbons message appears spoofed, dropping");
```

Comment:
I'd prefer to see a better INFO log here, this message does not provide the Openfire admin with any helpful info, like who the stanza is from or how it is making the spoof test.